src/backends/plonky2/primitives/merkletree_circuit.rs

```diff
@@ -0,0 +1,484 @@
+        // - existence: leaf contains the given key & value
+        // - non-existence:
+        //      - case i) expected leaf does not exist
+        //      - case ii) expected leaf does exist but it has a different key
+        //
+        // The following table expresses the options with their in-circuit
+        // selectors:
+        // | existence   | case_ii   | leaf_hash                    |
+        // | ----------- | --------- | ---------------------------- |
+        // | 1           | 0         | H(key, value, 1)             |
+        // | 0           | 0         | EMPTY_HASH                   |
+        // | 0           | 1         | H(other_key, other_value, 1) |
+        // | 1           | 1         | invalid combination          |
+
+        // First, ensure that both existence & case_ii are not true at the same
+        // time:
+        // 1. sum = existence + case_ii_selector
+        let sum = builder.add(existence.target, case_ii_selector.target);
+        // 2. sum * (sum-1) == 0
+        builder.assert_bool(BoolTarget::new_unsafe(sum));
+
+        // define the case_i_selector as true when both existence and
+        // case_ii_selector are false:
+        //     case_i_selector = (1 - existence) * (1 - case_ii_selector)
+        let one = builder.one();
+        let existence_inv = builder.sub(one, existence.target);
+        let case_ii_inv = builder.sub(one, case_ii_selector.target);
+        let case_i_selector = BoolTarget::new_unsafe(builder.mul(existence_inv, case_ii_inv));
+
+        // use (key,value) or (other_key, other_value) depending if it's a proof
+        // of existence or of non-existence, ie:
+        // k = key * existence + other_key * (1-existence)
+        // v = value * existence + other_value * (1-existence)
+        let k: Vec<Target> = (0..4)
+            .map(|j| builder.select(existence, key[j], other_key[j]))
+            .collect();
+        let v: Vec<Target> = (0..4)
+            .map(|j| builder.select(existence, value[j], other_value[j]))
+            .collect();
+
+        // get leaf's hash for the selected k & v
+        let h = Self::kv_hash_target(builder, &k, &v);
+
+        // if we're in the case i), use leaf_hash=EMPTY_HASH, else use the
+        // previously computed hash h.
+        let empty_hash = builder.constant_hash(HashOut::from(EMPTY_HASH.0));
+        let leaf_hash = HashOutTarget::from_vec(
+            (0..4)
+                .map(|j| builder.select(case_i_selector, empty_hash.elements[j], h.elements[j]))
+                .collect(),
+        );
+
+        // get key's path
+        let path = Self::keypath_target(builder, &key);
+
+        // compute the root for the given siblings and the computed leaf_hash
+        // (this is for the three cases (existence, non-existence case i, and
+        // non-existence case ii)
+        let computed_root = Self::compute_root_from_leaf(builder, &path, &leaf_hash, &siblings)?;
```

Comment:
This is a very small optimization.
Since the `computed_root` is equal to `root`, we can just call this `root` and remove some duplicated wires:

```suggestion
        let root = Self::compute_root_from_leaf(builder, &path, &leaf_hash, &siblings)?;
```
And remove this line from the beginning of the function:
```rust
        let root = builder.add_virtual_hash();
```

Reply:
Good point! updated

Reply:
I've just learned that a virtual target is not mapped to an actual circuit wire.  From the plonky2 code:
```rust
    /// Adds a new "virtual" target. This is not an actual wire in the witness, but just a target
    /// that help facilitate witness generation. In particular, a generator can assign a values to a
    /// virtual target, which can then be copied to other (virtual or concrete) targets. When we
    /// generate the final witness (a grid of wire values), these virtual targets will go away.
    pub fn add_virtual_target(&mut self) -> Target {
```

So my suggestion doesn't bring any optimization.  Nevertheless I still think it's useful to keep the change because it simplifies the code.